### ISSUE TITLE
Doors and Windows recognition

### ISSUE THREAD

**Issue by Tim-HW** [grebtsew/FloorplanToBlender3d#5]
Hello sir,

First, I would like to thank you for you code, It is awesome and it's saving me a lot of developpement time.
However, I would need to recognize doors and windows for my project. I would be glad to made a pull request once the enhancement finished if you want. But before starting I would need your advise on how to do it. 

I'll be glad to know you thought about it.

Kind Regards from Belgium

Tim

**Comment by grebtsew** [comment]
Hello @Tim-HW,

Thank you for checking out the implementation! It's great to hear that it's helpful.

I actually looked into adding door and window recognition about a year ago. However didn't have time to implement anything. But perhaps now that I've seen two issues on this matter I should try to prioritize some time for it. So I might be able to help implement something after the 14th of june.

However my conclusions from a year ago is that I could see two reasonable alternatives.

1. Train a supervised NN model on finding doors and windows. This together with some rounding of corner-detections would work great and not be too much of a challange to implement. 
 
An advantage of this strategy is that we won't have to limit the implementation to only work  for instance on certain image sizes or floorplan styles. If we would want to get a better performance for a larger varity of images we would simply update the dataset and retrain. Possibly have several models.

A disadvantage is that we then preferably would add a new dependency to for example Tensorflow/Keras/pyTorch/skLearn for handling the AI parts, because if I remember correctly openCV library does not support training of new models. There would also be the issue with provisioning models, as they might become too large to be shared on github (>100mb).

2. Use Template matching instead. This is the solution I would prefer. The idea is that we have a couple of images that we know will have a certain style. So assume all doors look the same but with different transforms. Template matching which is a function we can use in openCV library then lets us know where in the image the doors exist. See image below.

![bild](https://user-images.githubusercontent.com/8101164/120728967-4b4c2f80-c4de-11eb-9322-ef6d8eca3fbc.png)

 We can now use already existing code for finding all doors and windows in the image, functions are called findDetails or something similiar. See result below. We can now distinguish the doors from windows by looking at the results from the template matching. The template matching can also be used to get rotations and so on.

![bild](https://user-images.githubusercontent.com/8101164/120728935-3ec7d700-c4de-11eb-84fa-81a1c97a8314.png)

Advantages is that I already written some example code to get started. It is located somewhere in the development center folders. We won't have to add more python libraries.

Disadvantage the results in the beginning will probably be a bit off. But when you have fixed some recurring problem cases, I think it can be nice.

Anyway, that are some of my thoughts on the matter. Hope it helps somewhat!

Regarding pull request, I will gladly accept pull requests as long as the code runs.

Good luck with you addition! Get in touch if you have any more questions. 
I unfortunately won't have time to help much codewise for a couple of weeks.

Cheer,

@grebtsew

**Comment by Tim-HW** [comment]
Hi @grebtsew,

Wow Thank you very much for this explicit and detailed answer. I totally agree with you: The NN will be too overkill for this type of recognition as well as to heavy for Github. I see that you already have, implemented a start of a solution! That great I am quite new at OpenCV so this would help me a lot.

I would have to finish another project first, but as soon as it is finished, I will start to continue your implementation.
I'll let you know once I have some progress ^^

Cheers

@Tim-HW

**Comment by saraalyy** [comment]
how can I draw the windows as opening in the wall in blender ??

**Comment by grebtsew** [comment]
@saraalyy I can take a look at it!

**Comment by saraalyy** [comment]
@grebtsew Thanks I appreciate it

**Comment by saraalyy** [comment]
@grebtsew  Did you manage to differentiate between windows and doors in blender?

**Comment by grebtsew** [comment]
Hello @saraalyy 

Sorry, I have been a litte preoccupied. However started writing something last night that seems to be quite accurate.

Its basicly feature extraction with some linear algebra to achive this:

<img width="1229" alt="door-detections" src="https://user-images.githubusercontent.com/8101164/131515120-cc8ecddb-2c7d-4671-983c-d985f1e7b20b.PNG">

Then I use the reverse of the wall detection to get details from floorplan:

<img width="779" alt="details" src="https://user-images.githubusercontent.com/8101164/131515228-88dd47d6-dbc7-45d8-b9eb-3ee03f947b40.PNG">

If we then compare doors with detailed boxes and apply a bandpass filter we can get a somewhat solid result:

<img width="934" alt="doors_and_windows" src="https://user-images.githubusercontent.com/8101164/131515432-04ce91e5-46e5-40cf-b409-bbc7adad2245.PNG">

Which as we can see seems to work fine. 

There are a bunch of extra steps I had to solve, the code needs to be optimized alot. I also need to add it to the library and add it to the demo script to get it in 3d. Also tweek some calibrations and test on more floorplans. 
I also found alot of fixes I have to do in the current code so I won't add this to master/main branch for a couple of weeks. However I can hopefully finish some tests and add a test-branch for windows and doors that you can try out in the end of the week.

**Comment by iiskakov** [comment]
Henlo  @grebtsew,

Any updates on the issue? ^_^

**Comment by grebtsew** [comment]
Hi @iiskakov ,

Yes, actually. Last week I started with the implementation of doors and windows in blender.
It seems to be working to some degree. However I have a few problems I need to fix before pushing up a test branch. My plan is to finish that tonight. If I get stuck on something I can atleast send a picture and status update of how it looks.

I have had alot to do, as usual, so I spend approximately an evening per week on this project.  
It's great motivation for me to hear from you.

**Comment by iiskakov** [comment]
Oh wow ^^
It's nice to hear that!

I wish you a great coding sesh then.
You did a great job in general; the speed and accuracy of the recognition is great.

**Comment by grebtsew** [comment]
Hi,

Unfortunately I did not finish the update today. Here is a status update:
 Currently doors are a little maleplaced and wrongly scaled. Windows are almost fine. The new code is really messy, so that must be fixed aswell. Hopefully I can spend some hours on this during the weekend. 

Some demo images of current implementation:

<img width="1308" alt="doors windows" src="https://user-images.githubusercontent.com/8101164/135546556-2364b795-b221-4ae5-be3f-980ea6b72227.PNG">

<img width="1193" alt="door windows2" src="https://user-images.githubusercontent.com/8101164/135546922-ecb658e9-683e-48a4-864b-65afa21c02e8.PNG">

**Comment by grebtsew** [comment]
Hi again,

Time for a new update!

A new branch is now available [here](https://github.com/grebtsew/FloorplanToBlender3d/tree/doors-and-windows)!

New features include:
 
- Checkout the config.ini file. It does now contain more settings. To reset the config.ini file to default , just remove it and run the program again.

- Checkout the ftbl/const.py file. It does not contain many costumizable variables. Feel free to play around with them.

- Window and Doors detection added. This is still in alpha phase, it does not work well with all pictures. This will need several fixes before working as good as other parts of the implementation, precision wise. Some reasons for the inaccuracy are that we are only looking at one door model at this point, making example.png good but not the others. That is also on my TODO to fix so that users can add own models. Another requirement is that the background is white for all windows to detect properly.

- Auto rescale of too small or too large images added, to make precision better on all images. Also in alpha phase as it won't work if no walls could be found. Perhaps that is reasonable though.

- Remove noise from phone taken images added.

- Major code cleanup, restucture and removal of "magic numbers".

- More TODO:s found.

Upcoming next couple of weeks:
I have unfortunately not had time to update the README or other documentation that is on my TODO.
Next up is to create a bunch of tests for all functions.

**Comment by iiskakov** [comment]
Sorry for dumb question, but I saw that there are some examples of doors and windows included with the repo.

Am I wrong assuming that they are for OpenCV to calibrate? Can they be safely changed to fit specific project, with, let's say wider windows?

Again, thank you for your work, it's a great project!

**Comment by iiskakov** [comment]
By the way, I just found out that Blender also comes with handy Archimesh Add-on. 
Add-on helps to create windows and doors. 

It looks like it's a step away from current implementation, and a great visual improvement. 
Do you think implementing that is possible?
I would do that myself, but I think my OpenCV-fu and Blender-fu is still to weak for that.

^^

**Comment by grebtsew** [comment]
Hi, 

Regarding your first question, 

The calibration folder contains images that are used for calibrations. At this point that is used to calculate a preferred image size depending on the size of the floorplan.

The model in the door folder is used as a model to find similar doors in the floorplan. My plan is to let the user add their own models in that folder in future. At this point you can change the door image if your floorplan uses another type of door for instance.

At this point, window models are not used. But will probably be added in the future.

Regarding the second question:

The Archimesh add-on seems to contain several useful classes for handling relevant objects. 
So that might be interesting to look at in the future. Perhaps this should be a separate feature issue. 

Cheers,
@grebtsew

**Comment by grebtsew** [comment]
Hi!

Today I updated the main/master branch with Window and Doors detections.
They can still be inaccurate, I recommend try changing some constants in const and changing door models or changing input image size to alter results.

I will therefore close this issue now. Feel free to open new issues.

Cheers,
@grebtsew
